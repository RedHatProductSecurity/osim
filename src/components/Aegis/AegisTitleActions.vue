<script setup lang="ts">
import { computed } from 'vue';

import AegisActions from '@/components/Aegis/AegisActions.vue';

import type { UseAegisSuggestDescriptionReturn } from '@/composables/aegis/useAegisSuggestDescription';

const props = defineProps<{
  composable: UseAegisSuggestDescriptionReturn;
}>();

const {
  canShowTitleFeedback,
  canSuggest,
  details: suggestionDetails,
  hasAppliedTitleSuggestion,
  isSuggesting,
  revertTitle,
  sendTitleFeedback,
  suggestDescription,
} = props.composable;

defineExpose({
  isSuggesting,
});

const suggestionTooltip = computed(() => {
  if (!hasAppliedTitleSuggestion.value || !suggestionDetails.value) return 'Suggest Title via AEGIS-AI';
  const { confidence, explanation, title, tools_used } = suggestionDetails.value;
  const parts: string[] = [];
  if (title) parts.push(`Value: ${title}`);
  if (confidence != null && confidence !== '') parts.push(`Confidence: ${confidence}`);
  if (explanation) parts.push(`Explanation: ${explanation}`);
  if (tools_used && tools_used.length > 0) {
    parts.push(`Tools Used: ${tools_used.join(', ')}`);
  }
  return parts.join('\n');
});
</script>

<template>
  <AegisActions
    :canSuggest="canSuggest"
    :hasAppliedSuggestion="hasAppliedTitleSuggestion"
    :hasMultipleSuggestions="false"
    :isSuggesting="isSuggesting"
    :canShowFeedback="canShowTitleFeedback"
    :suggestions="[]"
    :selectedIndex="0"
    :tooltipText="suggestionTooltip"
    featureFlag="aiDescriptionSuggestions"
    @suggest="suggestDescription"
    @revert="revertTitle"
    @feedback="sendTitleFeedback"
  />
</template>
