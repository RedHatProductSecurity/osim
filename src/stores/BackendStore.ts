import { ref } from 'vue';
import { defineStore } from 'pinia';


export interface OsimBackends {
  osidb: string | unknown;
  isReady: () => boolean;
  error: boolean | string;
}

export const useBackendStore = defineStore('BackendStore', () => {
  const error = ref<string>('');
  const osidb = ref<string>('');

  function isReady() {
    return osidb.value !== '';
  }

  return {
    error,
    osidb,
    isReady,
  };
});
