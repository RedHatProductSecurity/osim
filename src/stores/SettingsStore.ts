import { ref, watch } from 'vue';
import { defineStore } from 'pinia';
import { z } from 'zod';

import serviceWorkerClient from '../services/service-worker-client';


// Consideration: we may want some settings to be persistently stored,
//   but other settings must be in-memory only, e.g. api keys

const _settingsStoreKey = 'SettingsStore';


export const SettingsSchema = z.object({
  // bugzillaApiKey: z.string().length(
  //   32, {message: 'Bugzilla API Key is the wrong length!'}
  // ).optional(),
  bugzillaApiKey: z.string().optional().or(z.literal('')),
  jiraApiKey: z.string().optional().or(z.literal('')),
  showNotifications: z.boolean(),
});
export type SettingsType = z.infer<typeof SettingsSchema>;

// const settingsStoreSessionStorage = z.object({
//   settings: SettingsSchema,
// });
// type SettingsStoreSessionStorage = z.infer<typeof settingsStoreSessionStorage>;


export const useSettingsStore = defineStore('SettingsStore', () => {
  const settings = ref<SettingsType>({ showNotifications: false });
  // const settings = useSessionStorage(_settingsStoreKey, {} as SettingsType);
  serviceWorkerClient.listen(_settingsStoreKey, value => {
    const newSettingsStore = SettingsSchema.safeParse(value);
    if (newSettingsStore.success) {
      if (JSON.stringify(newSettingsStore.data) !== JSON.stringify(settings.value)) {
        // New value; update
        settings.value = newSettingsStore.data;
      }
    }
  });
  watch(settings, () => {
    serviceWorkerClient.put(_settingsStoreKey, JSON.parse(JSON.stringify(settings.value)));
  });

  function save(newSettings: SettingsType) {
    settings.value = { ...newSettings };
  }

  function $reset() {
    settings.value = { showNotifications: false };
  }

  return {
    $reset,
    save,
    settings,
  };
});

