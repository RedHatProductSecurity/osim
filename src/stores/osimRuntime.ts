// No need to use Pinia; these values are set on app startup and never changed.
// Named with camelCase instead of PascalCase to differentiate from Pinia stores.

import { computed, ref } from 'vue';
import { z } from 'zod';

// export enum OsimRuntimeDev {
//   PROD,
//   DEV,
//   MOCK,
// }
// export const osimRuntimeDev = OsimRuntimeDev.DEV;

export enum OsimRuntimeStatus {
  INIT,
  READY,
  ERROR,
}

const status = ref(OsimRuntimeStatus.INIT);
export const osimRuntimeStatus = computed<OsimRuntimeStatus>(() => {
  return status.value;
});

const OsimRuntime = z.object({
  backends: z.object({
    osidb: z.string(),
    bugzilla: z.string(),
  }),
  osimVersion: z.string(),
  error: z.string().default(''),
});
type OsimRuntime = z.infer<typeof OsimRuntime>;

const runtime = ref<OsimRuntime>({
  backends: { osidb: '', bugzilla: '' },
  osimVersion: '',
  error: '',
});
export const osimRuntime = computed<OsimRuntime>(() => {
  return runtime.value;
});


const OsidbHealthy = z.object({
  env: z.string(),
  revision: z.string(),
  version: z.string(),
});
type OsidbHealthy = z.infer<typeof OsidbHealthy>;
const _osidbHealth = ref<OsidbHealthy>({
  env: '',
  revision: '',
  version: '',
});
export const osidbHealth = computed<OsidbHealthy>(() => {
  return _osidbHealth.value;
});



let setupCallCount = 0;
export function setup() {
  if (setupCallCount > 0) {
    console.warn('osimRuntime.setup called more than once:', setupCallCount + 1);
  }
  setupCallCount++;

  return fetchRuntime()
    .then(fetchOsidbHealthy)
    .then(() => {
      // If an error was not set, then the status is still INIT.
      // If the status is still INIT after all the initialization, then set the status to READY.
      if (status.value === OsimRuntimeStatus.INIT) {
        status.value = OsimRuntimeStatus.READY;
      }
    });
}

function fetchRuntime() {
  return fetch('/runtime.json', {
    method: 'GET',
    cache: 'no-cache',
  })
    .then(response => response.json())
    .then((json: OsimRuntime) => {
      console.debug('OsimRuntime', json);
      try {
        runtime.value = OsimRuntime.parse(json);
      } catch (e) {
        console.error('Unable to parse OsimRuntime', e);
        runtime.value.error = 'Backends are not correctly configured. Please try again later.';
        status.value = OsimRuntimeStatus.ERROR;
      }
    })
    .catch(e => {
      console.error('Unable to get backends', e);
      runtime.value.error = 'Error finding backends. ' +
                            'Possible deployment misconfiguration. Please try again.';
      status.value = OsimRuntimeStatus.ERROR;
    });

}

function fetchOsidbHealthy() {
  return fetch(`${runtime.value.backends.osidb}/osidb/healthy`, {
    method: 'GET',
    cache: 'no-cache',
  })
    .then(response => response.json())
    .then((json: OsidbHealthy) => {
      console.debug('OsidbHealthy', json);
      try {
        _osidbHealth.value = OsidbHealthy.parse(json);
      } catch (e) {
        console.error('Unable to parse OSIDB health', e);
        runtime.value.error = 'Unexpected response from OSIDB. Please try again later.';
        status.value = OsimRuntimeStatus.ERROR;
      }
    })
    .catch(e => {
      console.error('Unable to get OSIDB health', e);
      runtime.value.error = 'Error getting OSIDB health. ' +
                            'Possible deployment misconfiguration. Please try again.';
      status.value = OsimRuntimeStatus.ERROR;
    });
}
