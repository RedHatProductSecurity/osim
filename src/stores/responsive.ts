// Track header and footer for elements between to have responsive positioning

import { ref } from 'vue';

const navbarHeight = ref<number>(0);
const navbarBottom = ref<number>(0);

const footerTop = ref<number>(0);
const footerHeight = ref<number>(0);

export {
  navbarHeight,
  navbarBottom,
  footerTop,
  footerHeight,
};
