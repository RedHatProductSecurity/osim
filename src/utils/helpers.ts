// Source: https://github.com/vuejs/core/issues/5303#issuecomment-1543596383

import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );
