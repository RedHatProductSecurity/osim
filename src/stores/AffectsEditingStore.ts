import { ref, computed, onUnmounted, type Ref } from 'vue';

import { clone } from 'ramda';
import { defineStore } from 'pinia';

import { useFlaw } from '@/composables/useFlaw';
import { useFlawAffectsModel } from '@/composables/useFlawAffectsModel';

import type { ZodAffectType } from '@/types/zodAffect';
import { matcherForAffect } from '@/utils/helpers';

export const useAffectsEditingStore = defineStore('EditableAffects', () => {
  const { flaw } = useFlaw();

  const { initialAffects } = useFlawAffectsModel();
  const affectSelections = ref<ZodAffectType[]>([]);

  const {
    areAllAffectsSelectable,
    areAllAffectsSelected,
    isAffectSelected,
    isIndeterminateSelection,
    isSelectable,
    resetSelections,
    selectedAffects,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
  } = useAffectSelections(affectSelections, isBeingEdited);

  const affectValuesPriorEdit = ref<ZodAffectType[]>([]);
  const affectsBeingEdited = ref<ZodAffectType[]>([]);

  onUnmounted($reset);

  function isBeingEdited(affect: ZodAffectType) {
    const matchingAffect = affectsBeingEdited.value.find(matcherForAffect(affect)) as ZodAffectType;
    return affectsBeingEdited.value.includes(matchingAffect);
  }

  function getAffectPriorEdit(affect: ZodAffectType): ZodAffectType {
    return affectValuesPriorEdit.value.find(prior => prior.uuid && prior.uuid === affect.uuid) || affect;
  }

  function editAffect(affect: ZodAffectType) {
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }

    if (!isBeingEdited(affect)) {
      const affectToEdit = flaw.value.affects.find(matcherForAffect(affect));
      if (affectToEdit) {
        affectsBeingEdited.value.push(affectToEdit);
        const affectCopy = clone(affect) as ZodAffectType;
        affectValuesPriorEdit.value.push(affectCopy);
      }
    }
  }

  function commitChanges(affect: ZodAffectType) {
    const matchAffect = matcherForAffect(affect);
    const updateIndex = flaw.value.affects.findIndex(matchAffect);
    flaw.value.affects.splice(updateIndex, 1, affect);
    resetStagedAffectEdit(affect);
  }

  function cancelChanges(affect: ZodAffectType) {
    resetStagedAffectEdit(affect);
  }

  function revertAffectToLastSaved(affect: ZodAffectType) {
    const saved = initialAffects.value.find(a => a.uuid === affect.uuid);
    const index = flaw.value.affects.findIndex(a => a.uuid === affect.uuid);
    if (index !== -1 && saved) {
      flaw.value.affects[index] = { ...saved };
    }
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }
  }

  function resetStagedAffectEdit(affect: ZodAffectType) {
    const matchAffect = matcherForAffect(affect);
    const editingIndex = affectsBeingEdited.value.findIndex(matchAffect);
    affectsBeingEdited.value.splice(editingIndex, 1);
    affectValuesPriorEdit.value.splice(affectValuesPriorEdit.value.findIndex(matchAffect), 1);
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }
  }
  function editSelectedAffects() {
    selectedAffects.value.forEach(editAffect);
  }

  function commitAllChanges() {
    const affectsToCommit = [...affectsBeingEdited.value];
    affectsToCommit.forEach(commitChanges);
  }

  function cancelAllChanges() {
    const affectsToCancel = [...affectsBeingEdited.value];
    affectsToCancel.forEach(cancelChanges);
  }

  function $reset() {
    affectsBeingEdited.value = [];
    affectValuesPriorEdit.value = [];
  }

  return {
    cancelChanges,
    commitChanges,
    isBeingEdited,
    getAffectPriorEdit,
    editAffect,
    revertAffectToLastSaved,
    affectValuesPriorEdit,
    affectsBeingEdited,
    $reset,
    selectedAffects,
    areAllAffectsSelected,
    isIndeterminateSelection,
    isSelectable,
    areAllAffectsSelectable,
    isAffectSelected,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
    resetSelections,
    editSelectedAffects,
    commitAllChanges,
    cancelAllChanges,
  };
});

function useAffectSelections(selectedAffects: Ref<ZodAffectType[]>, isBeingEdited: (affect: ZodAffectType) => boolean) {
  const { flaw } = useFlaw();
  const affects = computed(() => flaw.value.affects);

  const { isAffectBeingRemoved } = useFlawAffectsModel();

  function isSelectable(affect: ZodAffectType) {
    return !isAffectBeingRemoved(affect) && !isBeingEdited(affect);
  }

  function isAffectSelected(affect: ZodAffectType) {
    return isSelectable(affect) && selectedAffects.value.includes(affect);
  }

  const areAllAffectsSelected = computed(() => {
    return affects.value.every(isAffectSelected);
  });

  const isIndeterminateSelection = computed(() => {
    return !areAllAffectsSelected.value && affects.value.some(isAffectSelected);
  });

  const areAllAffectsSelectable = computed(() => affects.value.every(isSelectable));

  function toggleAffectSelection(affect: ZodAffectType) {
    if (!isSelectable(affect)) {
      return;
    }
    if (!isAffectSelected(affect)) {
      selectedAffects.value.push(affect);
    } else {
      selectedAffects.value = selectedAffects.value.filter(a => a !== affect);
    }
  }

  function toggleMultipleAffectSelections(event: Event) {
    if (areAllAffectsSelected.value) {
      affects.value.forEach(affect => toggleAffectSelection(affect));
    } else if (selectedAffects.value.length === 0) {
      affects.value.filter(affect => !isAffectSelected(affect)).forEach(affect => toggleAffectSelection(affect));
    } else {
      // De-select all selected affects if only some are selected
      affects.value.filter(affect => isAffectSelected(affect)).forEach(toggleAffectSelection);
      (event.target as HTMLInputElement).checked = false;
    }
  }

  function resetSelections() {
    selectedAffects.value = [];
  }

  return {
    selectedAffects,
    areAllAffectsSelected,
    isIndeterminateSelection,
    isSelectable,
    areAllAffectsSelectable,
    isAffectSelected,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
    resetSelections,
  };
}
