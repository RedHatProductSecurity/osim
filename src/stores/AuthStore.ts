import { computed, ref, watch } from 'vue';

import { defineStore } from 'pinia';
import { z } from 'zod';
import jwtDecode from 'jwt-decode';
import type { JwtPayload } from 'jwt-decode';
import { useLocalStorage } from '@vueuse/core';
import { DateTime } from 'luxon';

import router from '@/router';
import { osimRuntime } from '@/stores/osimRuntime';

function getDevRefreshToken(): null | string {
  return osimRuntime.value.env === 'dev' ? (localStorage.getItem('AuthStore::refresh') || null) : null;
}

function setDevRefreshToken(token: string) {
  if (osimRuntime.value.env === 'dev') {
    localStorage.setItem('AuthStore::refresh', token);
  }
}

const loginResponse = z.object({
  access: z.string(),
  refresh: z.string().optional(),
  env: z.string(),
  detail: z.string().optional(),
});

export const useAuthStore = defineStore('AuthStore', () => {
  const accessToken = ref<null | string>();
  const refreshToken = ref<null | string>();

  // Persist only login state; tokens remain in-memory except dev refresh token (handled separately)
  const isLoggedIn = useLocalStorage('AuthStore::isLoggedIn', false);

  function $reset() {
    accessToken.value = null;
    refreshToken.value = null;
    isLoggedIn.value = false;
    // Clear dev refresh token storage
    localStorage.removeItem('AuthStore::refresh');
  }

  function isAccessTokenExpired() {
    try {
      const exp = accessToken.value ? jwtDecode<JwtPayload>(accessToken.value)?.exp : null;
      return !exp || DateTime.now().toSeconds() >= exp - 60;
    } catch (e) {
      console.debug('AuthStore: access token not a valid JWT', accessToken.value, e);
      return true;
    }
  }

  async function login(username: string = '', password: string = '') {
    const requestMetadata: { [key: string]: any } = {};
    if (osimRuntime.value.backends.osidbAuth == 'kerberos') {
      requestMetadata['method'] = 'GET';
    } else if (osimRuntime.value.backends.osidbAuth == 'credentials') {
      requestMetadata['method'] = 'POST';
      requestMetadata['body'] = JSON.stringify({ username: username, password: password });
      requestMetadata['headers'] = {
        'Content-Type': 'application/json',
      };
    }

    return fetch(`${osimRuntime.value.backends.osidb}/auth/token`, {
      credentials: 'include',
      cache: 'no-cache',
      ...requestMetadata,
    })
      .then(async (response) => {
        if (!response.ok) {
          const text = await response.text();
          throw new Error(text);
        }
        const json = await response.json();
        const parsedLoginResponse = loginResponse.passthrough().parse(json);
        if (parsedLoginResponse.detail) {
          throw new Error(parsedLoginResponse.detail);
        }
        // Store refresh token policy: dev in localStorage, non-dev in memory
        if (parsedLoginResponse.refresh) {
          if (osimRuntime.value.env === 'dev') {
            setDevRefreshToken(parsedLoginResponse.refresh);
            refreshToken.value = null;
          } else {
            refreshToken.value = parsedLoginResponse.refresh;
          }
        } else {
          refreshToken.value = null;
        }
        accessToken.value = parsedLoginResponse.access;
        isLoggedIn.value = true;
        // expose env to UserStore for persistence/compatibility
        const { useUserStore } = await import('./UserStore');
        useUserStore().setEnv(parsedLoginResponse.env);
        return parsedLoginResponse.access;
      })
      .then((access) => {
        // Fetch whoami and update user store without creating a static import cycle
        return fetch(`${osimRuntime.value.backends.osidb}/osidb/whoami`, {
          credentials: 'include',
          cache: 'no-cache',
          headers: {
            Authorization: `Bearer ${access}`,
          },
        });
      })
      .then(response => response.json())
      .then(async (json) => {
        const { whoamiResponse } = await import('./UserStore');
        const parsedWhoamiResponse = whoamiResponse.parse(json);
        const { useUserStore } = await import('./UserStore');
        const userStore = useUserStore();
        userStore.setWhoami(parsedWhoamiResponse);
      })
      .then(async () => {
        const { useUserStore } = await import('./UserStore');
        const userStore = useUserStore();
        await userStore.updateJiraUsername();
      })
      .catch((e) => {
        $reset();
        console.error('AuthStore::login() unsuccessful login request', e);
        throw e;
      });
  }

  function logout() {
    $reset();
    return router.push({ name: 'login' });
  }

  const isAuthenticated = computed<boolean>(() => {
    if (!accessToken.value) {
      return false;
    }
    try {
      const decoded = jwtDecode<JwtPayload>(accessToken.value);
      const now = Date.now() / 1000;
      return decoded.exp != null && now < decoded.exp;
    } catch (e) {
      return false;
    }
  });

  // React to authentication changes similarly to previous UserStore behavior
  watch(isAuthenticated, async () => {
    if (isAuthenticated.value) {
      if (router.currentRoute.value.name === 'login') {
        try {
          const { queryRedirect } = await import('./UserStore');
          const maybeRedirect = queryRedirect.parse(router.currentRoute.value);
          const redirect = maybeRedirect.query.redirect;
          if (redirect.startsWith('/')) {
            router.push(redirect);
            return;
          } else {
            console.debug('AuthStore::isAuthenticated() Refusing to redirect to', redirect);
          }
        } catch (e) {
          // do nothing
        }
        router.push({ name: 'index' });
        return;
      }
    } else {
      // Clear on unauthenticated and redirect to login
      $reset();
      if (router.currentRoute.value.name !== 'login') {
        const currentPath = router.currentRoute.value.fullPath;
        if (currentPath !== '/') {
          const query: any = {};
          query.redirect = currentPath;
          router.push({ name: 'login', query });
          return;
        }
        router.push({ name: 'login' });
        return;
      }
    }
  });

  return {
    isLoggedIn,
    accessToken,
    refreshToken,
    isAccessTokenExpired,
    isAuthenticated,
    getDevRefreshToken,
    setDevRefreshToken,
    login,
    logout,
    $reset,
  };
});
