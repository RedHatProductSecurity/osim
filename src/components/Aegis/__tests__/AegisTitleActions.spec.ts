import { computed } from 'vue';

import { describe, it, expect, vi, beforeEach } from 'vitest';

import AegisTitleActions from '@/components/Aegis/AegisTitleActions.vue';

import type { UseAegisSuggestDescriptionReturn } from '@/composables/aegis/useAegisSuggestDescription';

import { mountWithConfig } from '@/__tests__/helpers';
import { osimRuntime } from '@/stores/osimRuntime';

const createMockComposable = (overrides = {}): UseAegisSuggestDescriptionReturn => ({
  canShowTitleFeedback: computed(() => false),
  canShowDescriptionFeedback: computed(() => false),
  canSuggest: computed(() => true),
  details: computed(() => null),
  hasAppliedTitleSuggestion: computed(() => false),
  hasAppliedDescriptionSuggestion: computed(() => false),
  isSuggesting: computed(() => false),
  revertTitle: vi.fn(),
  revertDescription: vi.fn(),
  sendTitleFeedback: vi.fn(),
  sendDescriptionFeedback: vi.fn(),
  suggestDescription: vi.fn(),
  ...overrides,
} as UseAegisSuggestDescriptionReturn);

describe('aegisTitleActions', () => {
  beforeEach(() => {
    // @ts-expect-error - osimRuntime is readonly in tests
    osimRuntime.value = {
      ...osimRuntime.value,
      flags: {
        ...osimRuntime.value.flags,
        aiDescriptionSuggestions: true,
      },
    };
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const mockComposable = createMockComposable();
    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('uses AegisActions component', () => {
    const mockComposable = createMockComposable();
    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    expect(wrapper.findComponent({ name: 'AegisActions' }).exists()).toBe(true);
  });

  it('exposes isSuggesting state', () => {
    const mockComposable = createMockComposable();
    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    expect(wrapper.vm.isSuggesting).toBeDefined();
  });

  it('calls suggestDescription when suggest action is triggered', async () => {
    const mockSuggestDescription = vi.fn();
    const mockComposable = createMockComposable({
      suggestDescription: mockSuggestDescription,
    });

    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    const aegisActions = wrapper.findComponent({ name: 'AegisActions' });
    await aegisActions.vm.$emit('suggest');

    expect(mockSuggestDescription).toHaveBeenCalled();
  });

  it('calls revertTitle when revert action is triggered', async () => {
    const mockRevertTitle = vi.fn();
    const mockComposable = createMockComposable({
      canShowTitleFeedback: computed(() => true),
      revertTitle: mockRevertTitle,
    });

    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    const aegisActions = wrapper.findComponent({ name: 'AegisActions' });
    await aegisActions.vm.$emit('revert');

    expect(mockRevertTitle).toHaveBeenCalled();
  });

  it('calls sendTitleFeedback when feedback action is triggered', async () => {
    const mockSendTitleFeedback = vi.fn();
    const mockComposable = createMockComposable({
      canShowTitleFeedback: computed(() => true),
      sendTitleFeedback: mockSendTitleFeedback,
    });

    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    const aegisActions = wrapper.findComponent({ name: 'AegisActions' });
    await aegisActions.vm.$emit('feedback', 'positive');

    expect(mockSendTitleFeedback).toHaveBeenCalledWith('positive');
  });

  it('displays correct tooltip when suggestion is applied', () => {
    const mockComposable = createMockComposable({
      hasAppliedTitleSuggestion: computed(() => true),
      details: computed(() => ({
        title: 'Test Title',
        confidence: 0.95,
        explanation: 'Test explanation',
        tools_used: ['tool1', 'tool2'],
      })),
    });

    const wrapper = mountWithConfig(AegisTitleActions, {
      props: {
        composable: mockComposable,
      },
    });

    const aegisActions = wrapper.findComponent({ name: 'AegisActions' });
    const tooltip = aegisActions.props('tooltipText');

    expect(tooltip).toContain('Test Title');
    expect(tooltip).toContain('0.95');
    expect(tooltip).toContain('Test explanation');
    expect(tooltip).toContain('tool1, tool2');
  });
});
