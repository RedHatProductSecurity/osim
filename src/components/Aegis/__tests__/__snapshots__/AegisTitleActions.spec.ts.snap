// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`aegisTitleActions > renders correctly 1`] = `
"<div data-v-e049df0b="" class="d-flex align-items-center"><i data-v-e049df0b="" class="bi-stars label-icon" title="Suggest Title via AEGIS-AI"></i>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
